### PR TITLE
Feature/replace serial cmds

### PR DESCRIPTION
**Pull request by keeramis**
## TODOs

- [x] Update particle-usb to include getNetworkInterfaceList

## Description
-------------

### This PR does the following
1. Replaces usage of serial commands with control requests
2. Removes `particle setup`
3. Removes `particle serial claim`
4. Removes `particle device doctor`

-------------

### Serial commands replaced with control requests
The following serial commands are replaced.

1. `m` : Device's wifi mac address
         - This command gives the mac address of wifi and ethernet
         - image

5. `i` : Device identify
        - Device identity is obtained normally but through control requests
        - image
        - If a cellular device is on an older device-os version which does not support obtaining IMEI through a control requests, a blank is returned
        - image

6. `s`: Device inspect
       - A new format for firmware modules is introduced in dvos 5.6.0 which contains extensive info about modules as well as info about assets. 
       - image

-------------

## How to Test

1. Prepare a device with this branch of device-os https://github.com/particle-iot/device-os/pull/2730

Run the following commands on devices with different platforms

- `npm start -- serial identify`
- `npm start -- serial inspect`
- `npm start -- serial mac`

## Related Issues / Discussions

https://app.shortcut.com/particle/epic/124394

Device-OS : https://github.com/particle-iot/device-os/pull/2730
Device-os-protobuf : https://github.com/particle-iot/device-os-protobuf/pull/26
Particle-USB : https://github.com/particle-iot/device-os/pull/2730

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [ ] Docs have been updated
- [ ] CI is passing

### CODE REVIEW

File: src/cmd/serial.js [hunk 1]
```diff
@@ -233,133 +243,308 @@ module.exports = class SerialCommand extends CLICommandBase {
...
+			const features = deviceConstants[platform].features;
+
+			if (features.includes('cellular')) {
+				// Add checking for new dvos versions
+				let res;
+				if (semver.lt(fwVer, '5.6.0')) { // FIXME: Get the correct firmware version
+					res = await this._identifyCellularDeviceOverSerial(deviceFromSerialPort);
+				} else {
+					res = await this._identifyCellularDeviceOverCtrlReq(device);
 				}
```
Comment (monkbroc) [inline src/cmd/serial.js:284]: I'd prefer not to keep a part of the identify command using a serial one letter request. I'd rather omit information that is not available as long as it's not critical. In this case, I expect nobody cares about the IMEI.

File: src/cmd/serial.js [hunk 2]
```diff
@@ -233,133 +246,265 @@ module.exports = class SerialCommand extends CLICommandBase {
...
+			deviceFromSerialPort = await this.whatSerialPortDidYouMean(port, true);
+			deviceId = deviceFromSerialPort.deviceId;
+			device = await usbUtils.getOneUsbDevice({ idOrName: deviceId });
+		} catch (err) {
+			if (!deviceId) {
+				// The main step has failed and we dont continue
+				// because the others would fail as well
+				throw new VError(err, 'Could not identify device');
+			}
+		}
```
Comment (hugomontero) [inline src/cmd/serial.js:280]: Can you extract this `try/catch` sections into small methods? in that way you can have something like:

```
async identifyDevice({port}) {
    ...
   const  device  =  await pickDeviceByPort(port);
   const fw = await getSystemVersion(device);
    ....
}
```

File: src/cmd/serial.js [hunk 3]
```diff
@@ -233,133 +246,265 @@ module.exports = class SerialCommand extends CLICommandBase {
...
+	}
+	
+	/**
+	 * Obtains module info from control requests
+	 * @param {object} device
+	 * @returns {boolean} returns error or true
+	 */
+	async _getModuleInfo(device) {
+		// TODO: Get this from the protobuf exports
+		const fwModuleDisplayNames = {
```
Comment (hugomontero) [inline src/cmd/serial.js:443]: Idk if that's the meaning of your TODO (if so, just ignore this comment 😄 ) but there is in particle-usb an object that is exported with this data: https://github.com/particle-iot/particle-usb/blob/590d4e7caf43aedc8e14df3eeb1663d26d058905/src/device.js#L63C7-L63C33 you can use it instead

File: src/cmd/serial.js [hunk 4]
```diff
@@ -233,133 +246,265 @@ module.exports = class SerialCommand extends CLICommandBase {
...
+			'SETTINGS': 'Settings',
+			'NCP_FIRMWARE': 'Network Co-processor Firmware',
+			'RADIO_STACK': 'Radio Stack Module',
+			'ASSET': 'Asset'
+		};
 
-				if (d.p !== undefined){
-					const platformName = knownPlatformDisplayForId()[d.p];
-					console.log('Platform:', d.p, platformName ? ('- ' + chalk.bold.cyan(platformName)) : '');
+		const modules = await device.getFirmwareModuleInfo();
```
Comment (hugomontero) [inline src/cmd/serial.js:456]: This should be a TODO in particle-usb 😛 => This method doesn't have timeout, sometimes if the device is "busy" this call will last for long time hehe. Is not intentional that doesn't have timeout, when I created that function I didn't know how the things worked in particle-usb. So in summary: We need to add timeout as param to this function 😄

File: src/cmd/serial.js [hunk 5]
```diff
@@ -27,6 +28,15 @@ const arrow = chalk.green('>');
...
 const timeoutError = 'Serial timed out';
 
+const FW_MODULE_INTEGRITY_CHECK_FLAG = 0x02;
+const FW_MODULE_DEP_CHECK_FLAG = 0x04;
+const FW_MODULE_RANGE_CHECK_FLAG = 0x08;
+const FW_MODULE_PLATFORM_CHECK_FLAG = 0x10;
+
+const MAC_ADDR_SIZE_BYTES = 6;
+
+// const availability = (asset, availableAssets) => availableAssets.some(availableAsset => availableAsset.hash === asset.hash);
```
Comment (hugomontero) [inline src/cmd/serial.js:38]: is it intended to be removed?

File: assets/doctor.ino [hunk 6]
```diff
@@ -2,6 +2,8 @@
 // device settings like IP address and antenna
 // Compile by running scripts/build_doctor.sh
 
+// CCC: Do we still need the script after the removal of the doctor command?
```
Comment (monkbroc) [inline assets/doctor.ino:5]: Delete this and the doctor binaries when removing the doctor command

File: src/cmd/serial.js [hunk 7]
```diff
@@ -11,32 +11,28 @@ const SerialPort = require('../lib/require-optional')('serialport');
...
 
-function protip(){
-	const args = Array.prototype.slice.call(arguments);
-	args.unshift(chalk.cyan('!'), chalk.bold.white('PROTIP:'));
-	console.log.apply(null, args);
-}
+const FW_MODULE_INTEGRITY_CHECK_FLAG = 0x02;
+const FW_MODULE_DEP_CHECK_FLAG = 0x04;
+const FW_MODULE_RANGE_CHECK_FLAG = 0x08;
+const FW_MODULE_PLATFORM_CHECK_FLAG = 0x10;
```
Comment (keeramis) [inline src/cmd/serial.js:33]: Reference: https://github.com/particle-iot/device-os-protobuf/blob/a1e3ef4412fdd8ab4488a8017bcdfb3599327fdf/cloud/describe.proto#L34

**Review by monkbroc — APPROVED** [review]
Nice work! Let's ship it! :shipit:

**Pull request merged**